### ISSUE TITLE
feature request: unicode in source

### ISSUE THREAD

**Issue by tpapp** [stan-dev/stan#1888]
# Introduction

Some languages now support Unicode (mostly UTF8) for writing source code. It would be great if one could also use Unicode in Stan source. 

Broadly, there are two possible levels of support:
1. in variable and function names (eg `ϕ`), and
2. in operators (eg `≤`), which provide synonyms for existing ones (eg `<=`)

# Example
This is how the [8 schools](https://github.com/stan-dev/example-models/blob/master/misc/eight_schools/eight_schools.stan) example would look like in unicode:
```stan
data {
  int<lower=0> J;             // number of schools
  real y[J];                  // estimated treatment effect (school j)
  real<lower=0> σ[J];         // std err of effect estimate (school j)
}
parameters {
  real μ;
  real θ[J];
  real<lower=0> τ;
}
model {
  θ ~ normal(μ, τ); 
  y ~ normal(θ, σ);
}
```

# Possible benefits
1. more compact source code
2. better mapping to equations in papers

# Possible downsides
1. editor/entry support 
2. font support
3. possibly corrupted files

The first two are mitigated by the fact that ASCII is a subset of UTF8, so using the feature is optional.

# UTF8 support in various languages which have interfaces for Stan
| language | identifiers | operators |
| ---- | ---- | ---- |
| R | yes | no | 
| Python | only from version 3 | no |
| Julia | yes | yes |
| Matlab | yes, [but needs to be enabled](http://de.mathworks.com/matlabcentral/newsreader/view_thread/238995) | no |
| Stata | yes, [from version 14](http://blog.stata.com/tag/unicode/) | no |

# Editor support

## Emacs

See [this list](https://github.com/vspinu/math-symbol-lists) for various UTF8 implementations using autocomplete, company-mode, and quail.

# See also
- [discussion on stan-dev](https://groups.google.com/d/topic/stan-dev/fd76xNO8i20/discussion)
- [detailed description for Julia](http://docs.julialang.org/en/release-0.4/manual/variables/#allowed-variable-names)

**Comment by ariddell** [comment]
Python 2 handles UTF-8 fine. (It's just not the default representation.)

**Comment by tpapp** [comment]
@ariddell: Can you provide a link with an example/description? Then I could update the list.

**Comment by bgoodri** [comment]
RStudio should work https://support.rstudio.com/hc/en-us/articles/200532197-Character-Encoding

**Comment by ariddell** [comment]
@tpapp https://docs.python.org/2/howto/unicode.html Python 3 uses Unicode by default. In Python 2 you need to be explicit about it. In both cases you can have unicode strings in source code.

**Comment by tpapp** [comment]
@ariddell: in the page you link I could not find an example with unicode identifiers (only strings, literals, filenames, etc).

**Comment by ariddell** [comment]
Here's a link to the exact section: https://docs.python.org/2/howto/unicode.html#unicode-literals-in-python-source-code

**Comment by bob-carpenter** [comment]
Those are unicode literals, not unicode identifiers.

Can you have 

  éø = 10

where you assign to a unicode literal?  Or dictionaries
with unicode keys?

- Bob

**Comment by ariddell** [comment]
You're right. In Python 2.7 you can't have unicode variables. In Python 3 you can. But why does that matter? We only need unicode in the Stan program code. (Parameter lookups aren't affected since keys are and always were strings.)

**Comment by bob-carpenter** [comment]
It looks like you use a dictionary structure for variable
names.

schools_dat = {
  'J': 8,
  'y': [28,  8, -3,  7, -1,  1, 18, 12],
  'sigma': [15, 10, 16, 11,  9, 11, 10, 18]
}

Can the keys be unicode?

RStan can read data values out of the environment if they're
named after variables in the Stan program.  And it can attach
the resulting draws as variables in the environment.

- Bob

**Comment by ariddell** [comment]
Python 2 has no problems with unicode dictionary keys. In fact, it *can*
have unicode variables in the environment but you have to reference them
via strings indirectly. For example, this works in Python 2:

```
>>> locals()[u'é'] = 9
>>> locals()[u'é']
```

(``locals`` is something like ``baseenv`` or ``.GlobalEnv`` in R)

Bref, there is nothing Python 2 can't do that's relevant to supporting
unicode in Stan code. The table above is inaccurate.

**Comment by bob-carpenter** [comment]
> On May 12, 2016, at 8:20 AM, Allen Riddell <notifications@github.com> wrote:
> 
> Python 2 has no problems with unicode dictionary keys. In fact, it *can*
> have unicode variables in the environment but you have to reference them
> via strings indirectly. For example, this works in Python 2:
> 
> ```
> >>> locals()[u'é'] = 9
> >>> locals()[u'é']
> ```
> 
> (``locals`` is something like ``baseenv`` or ``.GlobalEnv`` in R)
> 
> Bref, there is nothing Python 2 can't do that's relevant to supporting
> unicode in Stan code. The table above is inaccurate.

You should have edit permission on the issues.

- Bob

**Comment by ariddell** [comment]
I was just recording my thought on the matter. I appreciate @tpapp putting work into drafting the issue text and would prefer to leave any edits to him.

**Comment by tpapp** [comment]
@ariddell: The table was accurate, but since not all Stan interfaces work the way that R/Julia does, I extended it with the information that is probably most relevant: whether the interfaces, in the way they currently operate, would support UTF8 variables for (1) passing data to Stan and (2) extracting MCMC results. Thanks for pointing this out, this is much more important than the details of UTF8 support in those languages *per se*.

Not being a STATA user, I am reluctant to make a definitive statement about it. If someone could help with that it would be great.

**Comment by ariddell** [comment]
One use of unicode in Stan Program code which should definitely be supported is in comments. Leaving code comments in one's native language is fairly routine in Python/Java/etc. We should at least support that in Stan.

**Comment by bob-carpenter** [comment]
Unicode in comments is OK now.

- Bob

**Comment by tpapp** [comment]
Indeed UTF8 comments work fine, and I have been using them for a while. Made a clarification in the issue.

**Comment by ariddell** [comment]
UTF8 comments aren't supported in PyStan right now (non-ASCII characters will generate an error). I'll fix this. pystan##223
